### ISSUE TITLE
[java] Members of record should be in scope in record header

### ISSUE THREAD

**Issue by jborgers** [pmd/pmd#5689]
**Affects PMD Version: 7.10, 7.12** 

**Description:**

Java XPath typeIs() function cannot resolve a type in certain situations, while it should be able to.  

**Code Samples demonstrating the issue:**

Example 1:
```java
import lombok.Builder;
import com.fasterxml.jackson.annotation.JsonFormat;
import java.time.LocalDateTime;

@Builder
public record UnresolvedTypeTryout (
            Long documentId,
            @JsonFormat(pattern = UnresolvedTypeTryout.DATE_TIME_PATTERN) // works fine, typeIs can resolve
            LocalDateTime requestedOn,
            @JsonFormat(pattern = DATE_TIME_PATTERN) // <-- typeIs can**not** resolve, while it should
            LocalDateTime refusedOn
) {
        private static final String DATE_TIME_PATTERN = "yyyy-MM-dd'T'HH:mm:ss";
}
```
Example 2:
```java
@Slf4j
 class EnvironmentLogService {
    public void init() {
        Environment.logEnv("server environment:", log); // typeIs can**not** resolve 'log', while it should
        unknownType.doIt(); // bad, typeIs cannot resolve and is correct
    }
}
class Environment {
    void logEnv(String text, Log log) {}
}
```

**Steps to reproduce:**

We created a custom rule to find places where types cannot be resolved, because then rules depending on typeIs() will not work correctly, and we want to know that.  This can be tested for instance in Designer.

**Running PMD through:** *Designer *

We created the rule with the following XPath:
```xpath
(: variable access for unknown variable type :)
(//VariableAccess
[not(pmd-java:typeIs('java.lang.Object'))]
[not(pmd-java:typeIs('byte'))][not(pmd-java:typeIs('short'))]
[not(pmd-java:typeIs('int'))][not(pmd-java:typeIs('long'))]
[not(pmd-java:typeIs('float'))][not(pmd-java:typeIs('double'))]
[not(pmd-java:typeIs('boolean'))][not(pmd-java:typeIs('char'))]
|
(: the unresolved/ambiguous type names :)
//AmbiguousName)
```

and to deal with the typeIs shortcomings:

```text
(: variable access for unknown variable type :)
(//VariableAccess
[not(pmd-java:typeIs('java.lang.Object'))]
[not(pmd-java:typeIs('byte'))][not(pmd-java:typeIs('short'))]
[not(pmd-java:typeIs('int'))][not(pmd-java:typeIs('long'))]
[not(pmd-java:typeIs('float'))][not(pmd-java:typeIs('double'))]
[not(pmd-java:typeIs('boolean'))][not(pmd-java:typeIs('char'))]
(: pmd seems not (always) able to resolve type of lambda params :)
[not(@Name = ancestor::LambdaExpression//LambdaParameter/VariableId/@Name)]
(: pmd seems not able to resolve record fields :)
[not(@Name = ancestor::RecordDeclaration/RecordBody/FieldDeclaration/VariableDeclarator/VariableId/@Name)]
|
(: the unresolved/ambiguous type names, except @Slf4j 'log' which pmd cannot resolve :)
//AmbiguousName)
[not(@Name='log' and
    ((ancestor::ClassDeclaration|ancestor::EnumDeclaration|ancestor::RecordDeclaration)/ModifierList/Annotation[pmd-java:typeIs('lombok.extern.slf4j.Slf4j')]))]
```

Please fix the typeIs() function.

**Comment by jborgers** [comment]
See https://github.com/jborgers/PMD-jPinpoint-rules/issues/581

**Comment by oowekyala** [comment]
The slf4j annotation is from Lombok. We could add support for it as part of #3119

**Comment by jborgers** [comment]
That would be nice. 

And it seems records are not fully supported, yet?
